File: go/schemapb/registry.go
package schemapb

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"sync"

	"google.golang.org/protobuf/proto"
)

// Registry errors. Implementations should return ErrNotFound from Get when a
// schema is absent.
var (
	ErrNotFound        = errors.New("schemapb: schema not found")
	ErrInvalidIdentity = errors.New("schemapb: schema identity requires a name")
)

// ListFilter narrows Registry.List. Empty fields are unconstrained; set
// fields must all match. NameContains is a case-insensitive substring match
// on the name. A nil *ListFilter matches everything.
type ListFilter struct {
	Namespace    string
	Name         string
	Version      string
	NameContains string
}

// Match reports whether id satisfies the filter.
func (f *ListFilter) Match(id *SchemaIdentity) bool {
	if f == nil {
		return true
	}
	if f.Namespace != "" && id.GetNamespace() != f.Namespace {
		return false
	}
	if f.Name != "" && id.GetName() != f.Name {
		return false
	}
	if f.Version != "" && id.GetVersion() != f.Version {
		return false
	}
	if f.NameContains != "" && !strings.Contains(strings.ToLower(id.GetName()), strings.ToLower(f.NameContains)) {
		return false
	}
	return true
}

// Registry stores schemas addressed by their identity. Implementations must
// be safe for concurrent use.
type Registry interface {
	// Put stores (or replaces) a schema under its own identity.
	Put(ctx context.Context, s *Schema) error
	// Get returns the schema for id, or ErrNotFound.
	Get(ctx context.Context, id *SchemaIdentity) (*Schema, error)
	// List returns the schemas matching filter (in unspecified order).
	List(ctx context.Context, filter *ListFilter) ([]*Schema, error)
}

// InMemoryRegistry is the default Registry: a concurrency-safe map. Nothing
// is persisted.
type InMemoryRegistry struct {
	mu sync.RWMutex
	m  map[string]*Schema
}

// NewInMemoryRegistry returns an empty in-memory registry.
func NewInMemoryRegistry() *InMemoryRegistry {
	return &InMemoryRegistry{m: map[string]*Schema{}}
}

// Put stores (or replaces) a schema under its own identity.
func (r *InMemoryRegistry) Put(_ context.Context, s *Schema) error {
	if s.GetId().GetName() == "" {
		return ErrInvalidIdentity
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	r.m[identityKey(s.GetId())] = s
	return nil
}

// Get returns the schema for id, or ErrNotFound.
func (r *InMemoryRegistry) Get(_ context.Context, id *SchemaIdentity) (*Schema, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	if s, ok := r.m[identityKey(id)]; ok {
		return s, nil
	}
	return nil, ErrNotFound
}

// List returns the schemas matching filter.
func (r *InMemoryRegistry) List(_ context.Context, f *ListFilter) ([]*Schema, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	out := make([]*Schema, 0, len(r.m))
	for _, s := range r.m {
		if f.Match(s.GetId()) {
			out = append(out, s)
		}
	}
	return out, nil
}

// =============================================================================
// Identity linking (reference composition)
// =============================================================================

// collectIDRefs walks fields recursively and records every identity-Ref by
// its identity key.
func collectIDRefs(fields []*Schema_Field, out map[string]*SchemaIdentity) {
	for _, f := range fields {
		if r := f.GetRef(); r != nil {
			if id := r.GetId(); id != nil {
				out[identityKey(id)] = id
			}
		}
		if l := f.GetList(); l != nil {
			collectIDRefs(l.GetItems(), out)
			continue
		}
		for _, child := range nestedSchemas(f) {
			collectIDRefs(child.GetFields(), out)
		}
	}
}

// Link resolves every identity-Ref in the schema against reg, pulling each
// referenced schema into the root defs (keyed by identity) so the schema
// becomes self-contained and validates standalone. The id-ref nodes keep
// their identity, so renderers can still see what each branch points at.
// Resolution is transitive: pulled-in schemas are themselves scanned for
// further id-refs. Link returns a linked clone; the receiver is not modified.
// An identity reg cannot supply, or a $defs key conflict, is an error.
func (s *Schema) Link(ctx context.Context, reg Registry) (*Schema, error) {
	root := proto.Clone(s).(*Schema)
	if root.Defs == nil {
		root.Defs = map[string]*Schema{}
	}
	if err := hoistDefs(root); err != nil {
		return nil, err
	}
	for {
		ids := map[string]*SchemaIdentity{}
		collectIDRefs(root.GetFields(), ids)
		for _, def := range root.Defs {
			collectIDRefs(def.GetFields(), ids)
		}
		added := false
		for key, id := range ids {
			if _, ok := root.Defs[key]; ok {
				continue
			}
			resolved, err := reg.Get(ctx, id)
			if err != nil {
				return nil, fmt.Errorf("schemapb: link: cannot resolve schema %s: %w", identityString(id), err)
			}
			clone := proto.Clone(resolved).(*Schema)
			for k, d := range clone.GetDefs() {
				if err := addDef(root, k, d); err != nil {
					return nil, err
				}
			}
			clone.Defs = nil
			root.Defs[key] = clone
			added = true
		}
		if !added {
			break
		}
	}
	return root, nil
}

File: py/src/schemapb/registry.py
"""Identity-addressed schema registry and identity-ref linking."""

from __future__ import annotations

import copy
from typing import TYPE_CHECKING

from schemapb.descriptor import nested_schemas

if TYPE_CHECKING:
    from schemapb._gen.schemapb import Schema, SchemaField, SchemaIdentity
    from schemapb.typed import Version


class RegistryError(Exception):
    """A registry misuse (missing name, conflicting identity, broken link)."""


def identity_key(id_: SchemaIdentity | None) -> str:
    if id_ is None:
        return "\x00\x00"
    return f"{id_.namespace}\x00{id_.name}\x00{id_.version}"


class InMemoryRegistry:
    """In-memory identity-addressed store; strict put + explicit replace."""

    def __init__(self) -> None:
        self._m: dict[str, Schema] = {}

    def put(self, s: Schema) -> None:
        if s.id is None or s.id.name == "":
            msg = "schemapb: schema identity requires a name"
            raise RegistryError(msg)
        key = identity_key(s.id)
        existing = self._m.get(key)
        if existing is not None and existing != s:
            msg = f"schemapb: identity already registered with different content: {s.id.name}"
            raise RegistryError(msg)
        self._m[key] = s

    def put_replace(self, s: Schema) -> None:
        if s.id is None or s.id.name == "":
            msg = "schemapb: schema identity requires a name"
            raise RegistryError(msg)
        self._m[identity_key(s.id)] = s

    def get(self, id_: SchemaIdentity) -> Schema | None:
        return self._m.get(identity_key(id_))

    def list(
        self,
        namespace: str | None = None,
        name: str | None = None,
        version: Version | None = None,
        name_contains: str | None = None,
    ) -> list[Schema]:
        out: list[Schema] = []
        for s in self._m.values():
            sid = s.id
            if namespace is not None and (sid is None or sid.namespace != namespace):
                continue
            if name is not None and (sid is None or sid.name != name):
                continue
            if version is not None and (sid is None or sid.version != str(version)):
                continue
            if name_contains is not None and (
                sid is None or name_contains.lower() not in sid.name.lower()
            ):
                continue
            out.append(s)
        return out


def _collect_id_refs(fields: list[SchemaField], out: dict[str, SchemaIdentity]) -> None:
    for f in fields:
        if f.ref is not None and f.ref.id is not None:
            out[identity_key(f.ref.id)] = f.ref.id
        if f.list is not None:
            _collect_id_refs(f.list.items, out)
            continue
        for child in nested_schemas(f):
            _collect_id_refs(child.fields, out)


def link(s: Schema, reg: InMemoryRegistry) -> Schema:
    """Pulls every identity-ref into the root defs (transitively); returns a
    linked deep copy.
    """
    root = copy.deepcopy(s)
    while True:
        ids: dict[str, SchemaIdentity] = {}
        _collect_id_refs(root.fields, ids)
        for def_ in root.defs.values():
            _collect_id_refs(def_.fields, ids)
        added = False
        for key, id_ in ids.items():
            if key in root.defs:
                continue
            resolved = reg.get(id_)
            if resolved is None:
                msg = f"schemapb: link: cannot resolve schema {id_.name}"
                raise RegistryError(msg)
            cloned = copy.deepcopy(resolved)
            for k, d in cloned.defs.items():
                root.defs.setdefault(k, d)
            cloned.defs = {}
            root.defs[key] = cloned
            added = True
        if not added:
            break
    return root
